\begin{savequote}[75mm] 
Knowing is not enough; we must apply. Willing is not enough; we must do
\qauthor{Goethe} 
\end{savequote}

\chapter{Use Cases} \label{section:use_cases}

\newthought{PINV has been used in several research projects}. The following are the study cases that we are most familiar with, and although the research in the presented cases is not part of this doctorate, we consider that it is relevant to introduce them to illustrate how PINV can be used to explore and visualise heterogeneous datasets. 

The cases below also serve to trace the evolution that the tool has experienced since its conception. The first two use cases listed below were research projects carried out in the same group where PINV was developed, the first of which exposed the need for a tool such as PINV and serve to defined the original requirements of PINV. With the second research project, these requirements were polished and finalised.  The third research project mentioned below is the first independent research that used PINV to explore their data and visualise their results. We got direct feedback from users, which helped us to resolve many bugs and to implement new features. The use of PINV's collaborative functions proved to be vital with the progression of this international collaboration. Lastly we present a study where the volume of data was bigger than previous datasets, which was used to test some of the server capabilities, and it is used here to show the potential of the spatial clustering technique.

\section{Predicting and Analyzing Interactions between \emph{Mycobacterium tuberculosis} and Its Human Host} \label{sec:mb_human}
\begin{figure}
\centering
\includegraphics[width=5in]{figures/pinv_human_mtb.png}
\caption[Predicted functional interactions between MTB and Human]{Predicted functional interactions between MTB and Human. Blue circles are human proteins while red squares are from the MTB microorganism.
\label{fig:pinv_human_mtb}}
\end{figure}

\subsection{Research description}
This research is based on the fact that diseases such as tuberculosis are better understood from the relationships between the pathogens and their host. Some of these relationships can be described as a protein protein interactions, where one protein belongs to the microorganism and the other to the host. Such interactions are not required to be direct (i.e proteins touching each other), and can be what is known as a functional interaction, where, for instance,  the presence of one protein has an effect on the function of another protein. 

Although highly relevant to the understanding of several biological processes, experimental studies of host-pathogen protein interactions are very scarce, and therefore computational prediction is the available alternative. In this study, the authors used the interologs method in order to predict functional interactions between\emph{Mycobacterium tuberculosis} MTB and \emph{Homo Sapiens}. The predicted interactions were filtered by selecting only the ones differently expressed during infection. The result set was then validated based in the known location of the proteins in the cell, verifying that they belong to an area where cross-organism interaction is actually possible. 

The 190 proteins found were further analysed using functional and metabolical pathway knowledge, confirming the potential for some of these proteins to be drug targets \cite{RAP2013}. These interactions also illustrate how MTB might acquire nutrients and how it modulates the host response to its advantage.

Figure \ref{fig:pinv_human_mtb} displays the interactions found in this research project using PINV.


\subsection{Impact on PINV}

When the first results of this research were presented in a group meeting, the authors expressed the difficulties experienced while exploring the interaction datasets, especially when trying to generate visualisations that would help them to explain their findings. Given our experience with the visualisation projects described in section \ref{section:dasvisual} we took the challenge of developing a tool that can support research projects that deal with this type of data on the web. 

Periodical meetings with the authors of this research helped us to define the requirements of PINV and other methods to improve the user experience of researchers dealing with protein-protein interaction data. 

The first observation from those meetings was that PINV needed to be more an exploratory tool than a graphic generator or an analysis application. And from there came the decision to have a repository that can be queried in order to extract only the information of interest, instead of loading the whole network in the visualiser, which ultimately would require the user to preprocess the data with a separate tool.

\begin{figure}
\centering
\includegraphics[width=4in]{figures/pinv_modes_query.png}
\caption[Three modes to query by ID in PINV]{Three modes to query by ID in PINV: Explicit, Normal and Recursive
\label{fig:pinv_modes_query}}
\end{figure}

It was with help of our collaborators that we defined the three modes to query by ID (Figure \ref{fig:pinv_modes_query}). PINV's firsts versions only included the normal mode in which a queried protein ID retrieves all the interactions where that protein ID has been reported, but later was pointed out that researchers some times prefer to limit their view to only a subset of proteins. We therefore developed the explicit mode to satisfy this requirement, and later the recursive mode was also added as a combination of the previous two options, allowing the user to obtain all the neighbours of a protein together with the interactions between them.

Another outcome from these meetings was related to the data itself: on the one hand, most of the interaction datasets have similarities -  interaction data is  limited to the definition of pairs of biological molecules, and information about the origin of the interaction is usually restricted to the method used to discover it. On the other hand, the protein data its heterogeneous - each project selects different features to consider, for some, network intrinsic annotations (e.g. degree, closeness, etc.) are the features to consider, while for others, biological features are more relevant (e.g. functional class, cellular location, etc.). 

Based on this, we determined that the upload of data to PINV requires two files: one defining the network with all the interactions between proteins and a second file with the annotations of interest for those proteins. A consequence of this open schema for protein data is that PINV widgets had to adapt to the specificities of each dataset. For example the definition of rules to manipulate the graphic representation considers the features defined for each project, figure \ref{fig:pinv_human_mtb} for instance, uses the degree (i.e. number of connections in the network) of each protein to resize it. Thanks to this, it is easy to identify highly connected proteins, even when the current visualisation is filtering out most of the other connections.

One more requirement gathered at this stage was to support direct manipulation of the visualisation: layouts are of great help to start analysing a network, however the researcher often needs to reorganise the network and manually locate some proteins in order to make  certain characteristics in the graphic more evident. Therefore, proteins in PINV can be moved by dragging them around the graphic, and the simulation forces can be stopped at any time.


\section{Orthologs in 3 Mycobactherium Organisms}\label{sec:orthologs}
\subsection{Research description}
This project presented a comparison between the functional networks of three related microorganisms: \emph{Mycobacterium tuberculosis} MTB, \emph{Mycobacterium leprae} MLP and \emph{Mycobacterium smegmatis}. Although sharing common ancestry, there are clear differences between them, some of which have been summarised in table \ref{tab:orthologs}.


\begin{table}[!ht]
        \begin{tabular}{|p{6cm}|p{3cm}|p{3cm}|p{3cm}|}
\hline 
\emph{feature} & \emph{MTB} & \emph{MLP} & \emph{MSM}\\
\hline 
Disease & Tuberculosis & Leprosy & \emph{non-pathogenic}\\
\hline 
Genome size & 4,411,532 & 3,268,203 & 6,988,209\\
\hline 
Number of proteins & 4136 & 1412 & 4853\\
\hline 
Doubling period (hours) & 24  & 500-670 & 3-4 \\
\hline 
        \end{tabular}
        \caption{Benchmarking between the main DAS servers.}
        \label{tab:orthologs}
\end{table}

The study identified the functional networks of the three organisms. The networks were created using STRING interactions as the base, but the network was enriched by including ten other sources of functional interactions. One of them was a method designed as part of the project in order to be able to use co-expression analysis for MLP proteins, because the number of available experiments in this organism was limited to four.

The final networks include only those interactions whose combined score ranges between medium or high confidence values, or those with low values that are present in two or more sources.  The network properties of these datasets were then analysed, the found values and further discussion about them can be found in \cite{AKI2013}.

The second part of the analysis compared the networks by pairs, in order to discover the biological consequences of the differences in the network. The strategy there was to select the most important proteins using network metrics (e.g. betweenness, degree and closeness) and proceed to comparing them with their orthologs in the other networks. The comparison was done using among other features the functional class of the proteins. 

Sub-networks were then created with only the proteins that have orthologs in the other two organisms, and similar comparisons to those explained above were carried out. Figure \ref{fig:pinv_orthologs} show an example of the type of analysis that can be done with the created networks. To start with, the authors have chosen one of the important proteins from MTB together with its orthologs in the other organisms, and by checking the neighbours of each protein and checking again for reported orthologs, it is easy to identify which interactions have been lost from one species to another.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/pinv_orthologs.png}
\caption[Three orthologs proteins in MTB, MLP and MSM]{Three orthologs proteins in MTB, MLP and MSM: Q7D903, Q9CD64, A0R3F9. and their neighbours.
\label{fig:pinv_orthologs}}
\end{figure}

The interactive version of figure \ref{fig:pinv_orthologs} can be seen at \url{http://biosual.cbio.uct.ac.za/pinViewer.html?status=a512969c93f31b50926415e7fec14096.json}. This visualisation includes the different images used in the published article of this research \cite{AKI2013}, all of which were created using PINV.

\subsection{Impact on PINV}
This project was also carried out in the same research group where PINV was developed and the projects were executed simultaneously, which again allowed us to have first hand contact with the potential users of PINV while still in design stages. As a consequence, some of the visualisation needs of this project influenced PINV development.

Firstly, this project required the display of three different species at a time, which made it evident that PINV should support the visualisation of interactions from multiple organisms. The  solution implemented in PINV was to visualise the proteins in the network using different SVG symbols for each organism,  meaning that PINV will automatically choose one symbol from those available when a protein of an organism that has not yet been displayed is added. Currently there are six available symbols, but the list can be extended in future. Figure \ref{fig:pinv_orthologs} uses circles for MTB, squares for MLP and diamonds for MSM.

As a complement, we extended the force-directed layout, including selective gravity forces, attracting each organism's proteins to different points in the graphic, which helps to automatically separate the proteins and avoids cluttered areas of multiple species.

One characteristic specific to this project was the use of information of orthologs in between species. We decided to include it in the dataset to be uploaded to PINV as another type of interaction, creating an extra evidence score column in the data. This allows the user to reuse all the functionalities defined for interactions. For example, the spring forces in the network view work the same with this pseudo-interaction, and as a result the orthologs are pulled out of the main group interactions of each organism. In figure \ref{fig:pinv_orthologs} for instance, the proteins that share orthologs in the other organisms have been pulled to the centre of the graphic, so they are easily identifiable.

\begin{figure}
\centering
\includegraphics[width=5in]{figures/pinv2dasty.png}
\caption[Opening DASty3 from a protein of interest in PINV.]{Opening DASty3 from a protein of interest in PINV.
\label{fig:pinv2dasty}}
\end{figure}

Another requirement gathered during this collaboration was the need to link the proteins of interest to other resources outside the PINV dataset, and being able to gain access to more information. Figure \ref{fig:pinv2dasty} shows on the left hand side the popup window that is displayed when a protein is selected in PINV, and includes the features that were uploaded to the dataset by the user. In addition, PINV also shows links to several external resources at the bottom of the popup window, and for instance if the user clicks on the DASty3 logo, this application will open in another window with related information on the selected protein.

\section{Shotgun Analysis of Platelets From Dengue Patients}
\label{sec:dengue}
\subsection{Research description}
In an attempt to understand better the phatogenesis of the dengue virus infection, a group of researchers from the Brazilian foundation Fiocruz have carried on a study to consider the impact of platelets during a dengue infection. Their strategy was to do a healthy v.s. infected quantification experiment using label free mass spectrometry.

The article with the details of this research is in the publication pipeline. Here we describe some of the preliminary results and the techniques used during the project: More than three thousand proteins were identified in two conditions: control and dengue. From those, the project identified 619 proteins that are either in only one of the two condition or that are differentially expressed, and there are 167 proteins that are present in both set of samples with differentially expressed profiles.

The identified proteins were then analysed in the context of their metabolic pathways by including information from the KEGG database. Similarly, the gene ontology (GO) was also considered in order to enrich the dataset with the biological functions.  A PPI network dataset was built using the interactions between the identified proteins which have been reported in the STRING database.

In a first interaction analysis, all the differentially express proteins and the proteins detected in only one condition were included. Figure \ref{fig:pinv_platelets_1} is representation of the network, which has been colour coded to easily identify the class of differential expression: light blue proteins were only detected in dengue patients and yellow represents the ones only identified in the control patients, the rest of the proteins were present in patients of both conditions, but red shows the proteins with under expression on dengue samples, while green proteins were over expressed.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/pinv_platelets_1.png}
\caption[PPI network of differentially expressed platelets proteins in a dengue study.]{PPI network of differentially expressed platelets proteins in a dengue study.
\label{fig:pinv_platelets_1}}
\end{figure}

When the functional classes from GO were considered, it was found that the eight most statiscally significant biological processes were related to “antigen processing and presentation” pathways, which drew the attention of the authors in context of the dengue pathological state. Some of these proteins are also related to proteasome/ubiquitinylation pathways.

To confirm whether these proteins directly interact with proteins related to the ``platelet activation'' pathway, a new graphic was made focusing on the specific non-redundant proteins involved in the ontology terms of interest. Figure \ref{fig:pinv_platelets_2} shows the subnetwork of interactions were at least one of the proteins were annotated with GO terms that contain the text: ``proteasome'', ``antigen'', ``platelet'' or ``mithocond''; colour coded in the figure by yellow, red, blue and pink respectively.

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/pinv_platelets_2.png}
\caption[Subnetwork of the differential expressed proteins identifying the most predominant GO terms.]{Subnetwork of the differential expressed proteins in the Dengue platelets dataset, identifying the most predominant GO terms by colour.
\label{fig:pinv_platelets_2}}
\end{figure}

\subsection{Impact on PINV}
We collaborated with the group responsible for this research in the creation of the network files and we also assist them to select an appropriate visualisation of their findings using PINV. The network was created using the reported interaction in STRING, which we filtered to include only the interactions in between the proteins of interest, i.e. those which have been identified in the mass spectrometry experiment, this required several preprocessing including mapping of accession numbers between resources and the enrichment of the existing annotations with gene ontology data.

We have documented the whole process hoping it will help future researchers with similar requirements. The git repository in \url{https://github.com/4ndr01d3/pinv-server/tree/master/data-preparation} contain the script used to create the files and explanations of the format of the required files and the logic behind the script.

The figures above were created by using PINV and the web versions of them are available at \url{http://biosual.cbio.uct.ac.za/pinViewer.html?status=67049496bf99027c6d6f3ba33a9cb77b.json} for figure \ref{fig:pinv_platelets_1} and \url{http://biosual.cbio.uct.ac.za/pinViewer.html?status=c1f68787907a83e27fab70be329582ce.json} for figure \ref{fig:pinv_platelets_2}.

The selection of these visualisations was a common effort between researchers of both groups, in which our peers from the Brazilian laboratory lead the process because they were familiar with the data and have the knowledge behind the dengue research, and we contributed with our experience in visualisation and the knowhow on PINV functionality.

The collaborative features of PINV demonstrated to be of great value in the communication between international groups. PINV as a native web tool does not require installation of any software besides a modern browser. Moreover, PINV capability to reconstruct s view from a generated URL supported the communication between the parts, allowing to point out the features that one group considered either important or irrelevant on a visualisation.

An example of the benefits of the visualisation of these networks with PINV was exposed during this project: on a previous version of the figure \ref{fig:pinv_platelets_2}, we were not considering the GO terms related with mithocondrial processes, however when exploring the visualisation was pointed out for an external collaborator about the presence of several proteins with different GO terms related with mitochondrial processes (e.g.  mitochondrion, mitochondrial inner membrane, positive regulation of mitochondrial depolarization, mitochondrial matrix), which gave another insight to the research and its biological relevance is still under consideration.



\section{Analysis of potential impact of SNPs on the human PPI network and gene expression}
\label{sec:pop_genetics}
\subsection{Research description}
This study investigates the impact of single nucleotide polymorphisms (SNP) on human particularly in the context of its protein-protein interaction network and taking into consideration gene expression for eight HapMap populations.

Previous studies have analysed consequences of SNPs in diseases, and some of them have included analysis to establish the relevance of the protein in which the mutation gets reflected in the human PPI network. Their finding suggests there are diseases that are caused by disrupted interactions. In \cite{HEE2014} the authors extend these studies in order to consider SNPs that are located in non-coding regions and the differences in allele frequencies between populations.

The project's methodology starts by integrating data from multiple resources of public domain: (i) human SNPs from NCBI (\url{ftp://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b141_GRCh37p13/VCF}), (ii) clinical related variations from NCBI (\url{http://www.ncbi.nlm.nih.gov/variation/docs/human_variation_vcf/#clinvar}) and Humsavar (\url{http://www.uniprot.org/docs/humsavar}), (ii) population data for allele frequencies (AF) and linkage disequilibrium (LD) for eight populations from HapMap (\url{ftp://ftp.ncbi.nlm.nih.gov/hapmap}), (iv) population gene expression data for the selected populations from Array Express (\url{http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-264/}), (v) SNPs associated with gene expression levels reported in the publications \cite{STR2007}, \cite{LAP2013}, \cite{HAU2014} and \cite{WU2013}, and (vi) the Human PPI network compiled in the research described in the article \cite{RAP2013}.

Three centrality measurements were used to define the importance of each protein in the network: degree, betweenness and closeness. When analysing these values for proteins with non-synonymous SNP (nsSNP) in the human PPI network, the data showed that proteins containing nsSNPs have significantly lower centrality measures than proteins without nsSNPs. This is an indication that proteins with less interactions are less important for information flow, and therefore they can tolerate the amino acid change without having a major impact in the network.

A similar centrality analysis was executed with the subset of proteins that contain SNPs that have been identified as linked to diseases, and in this case it was shown that these proteins tended to have significant higher centrality measures than proteins without. Which can be explained by the implications on the structure changes caused by the mutation, which if it is considerable, might inhibit some of the stablished interactions. This is not  critical for proteins with an small number of interactions, but for central proteins the effect can escalate to the level of collapsing the network, and hence the disease.

The integrated dataset can be used to explore particular diseases and potentially improve understanding of inter-population variation. An example of it was included in the research, in which proteins with SNPs related to hypertension were selected: ADD1, NOS3, AGT and KCNMB1. Figure \ref{fig:pinv_human_snps} shows the subnetwork of these proteins only displaying high confidence interactions (score ≥ 0.7) in which both of the proteins contained a nsSNP and at least one of the proteins contained a clinSNP associated with hypertension. The figure is colour coded to highlight the significant differences in AF and gene expression between African populations. 

\begin{figure}
\centering
\includegraphics[width=\textwidth]{figures/pinv_human_snps.png}
\caption[Interactions between proteins with a hypertension clinSNP]{Interactions between proteins with a hypertension clinSNP and proteins with nsSNPs in the context of AF and gene expression differences between African populations.
\label{fig:pinv_human_snps}}
\end{figure}

Similar comparison among other populations are included in \cite{HEE2014} discussing the implications of the identified proteins in more detail. All the graphics were created using PINV. An interactive version of Figure \ref{fig:pinv_human_snps} is available in \url{http://biosual.cbio.uct.ac.za/pinViewer.html?status=dc029e8ef629cd4fed0c4a1509a70ef7.json}

\subsection{Impact on PINV}
The case explained above where hypertension related proteins were analysed on the context of PPI networks was completely carried on using PINV. The dataset of this research was uploaded to the PINV server and is available under this URL: \url{http://biosual.cbio.uct.ac.za/pinViewer.html?core=HumanHapMapSNPs}. 

The authors were able to explore their data by defining different combinations of filters, for example, figure \ref{fig:pinv_prefilters_snp} is a snapshot of the selected filters used for the generation of figure \ref{fig:pinv_human_snps}, which we will describe from top to bottom: 
\begin{enumerate}
\setlength\itemsep{-0.5em}
\item This filters the whole human PPI network (1298519 interactions) in order to only include the ones with high confidence interactions (score ≥ 0.7), which is a total of 306453 interactions.
\item The second filter helps the researchers to focus only in the interactions between proteins with nsSNP. There are 182755 interactions in the whole network complying with this restriction, and in combination with the first filter, the subset gets reduced to 50211 interactions.
\item There is a total of 209540 interactions where at least one of the proteins has a SNP linked to a disease, and from that 9662 interactions also pass the previous filters.
\item This filter selects all the interactions where at least one of the proteins contains the term ``\emph{hypertension}'' on a field were all the diseases related with SNPs present in a protein are listed. 1711 interactions were found this way, and the intersection with the other filters produces 212 interactions.
\end{enumerate}
\begin{figure}
\centering
\includegraphics[width=5in]{figures/pinv_prefilters_snp.png}
\caption[Selection of filters for image \ref{fig:pinv_human_snps}]{Selection of filters for image \ref{fig:pinv_human_snps}
\label{fig:pinv_prefilters_snp}}
\end{figure}

From the result graphic only 4 proteins have been directly identified to be related with hypertension, and although it wasn't  imposed by the filters, the result subnetwork connects 3 of the proteins with one degree of separation in between them. Which agrees with the idea some diseases can be better understood by studying their related proteins in context of the human PPI network.

Once the network is identified, other factors can be analysed, for instance the colour coding in figure \ref{fig:pinv_human_snps} identifies differently expressed proteins in different african populations. This can help to understand why a disease affects populations in different ways. It is now part of our future plan to define a strategy to compare networks that share the same structure, and for example been able to see in the same graphic a comparison between the analysis for african populations \emph{v.s.} european or american populations. 

The example presented in this research demonstrates how PINV can be used for the exploration of PPI data, and how the use of filters supplies an alternative to navigate the data when the accession numbers of the proteins of interest are still to be found. A similar protocol to the described above can be use to explore other clinical conditions, for instance one of the appendices of \cite{HEE2014} is a similar visualisation but for cancer related proteins. This is an example of a dataset that by being public in PINV can benefit more than one research.

While the project was being carried on, the authors gave us constant feedback and many suggestions to improve PINV. For instance, several issues were corrected in the uploading process because of the large number of interactions that needed to be upload.

%to show the clustering in action!! 


